File: kotlin/uno/app/src/main/java/com/example/unonew/modules/LongestSubstringWithoutRepeatingCharacters.kt
package com.example.unonew.modules

class LongestSubstringWithoutRepeatingCharacters {
    fun lengthOfLongestSubstring(s: String): Int {
        val map = HashMap<Char, Int>()
        var start = 0
        var end = 0
        var max = 0
        while (end < s.length) {
            val current = s[end]
            val index = map[current]
            index?.let {
                start = Math.max(index, start)
            }
            max = Math.max(max, end - start + 1)
            map[current] = end + 1
            end++
        }
        return max
    }
    fun lengthOfLongestSubstring2(s: String): Int {
        val queue = LinkedList<Char>()
        var result = 0
        s.forEach {
            if (queue.contains(it)) {
                result = Math.max(queue.size, result)
                while (queue.pollFirst() != it) {
                    println(it)
                }
                queue.add(it)
            } else {
                queue.add(it)
            }
            result = Math.max(queue.size, result)
        }
        return result
    }
}

File: cmasmas/uno/LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.cpp
#include "LongestSubstringWithoutRepeatingCharacters.h"

class LongestSubstringWithoutRepeatingCharacters
{
public:
    int lengthOfLongestSubstring(std::string s)
    {
        int max = 0;
        int start = 0;
        int end = 0;
        int len = s.length();
        int i = 0;
        int j = 0;
        int index[256];
        for (i = 0; i < 256; i++)
        {
            index[i] = -1;
        }
        for (i = 0; i < len; i++)
        {
            j = s[i];
            if (index[j] >= start)
            {
                end = i;
                if (end - start > max)
                {
                    max = end - start;
                }
                start = index[j] + 1;
            }
            index[j] = i;
        }
        end = i;
        if (end - start > max)
        {
            max = end - start;
        }
        return max;
    }
};